File: backend/src/test/java/dev/coldhands/pair/stairs/backend/AbstractAcceptanceTest.java
package dev.coldhands.pair.stairs.backend;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.coldhands.pair.stairs.backend.domain.*;
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CalculateInputDto;
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.SaveCombinationEventDto;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import java.time.LocalDate;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public abstract class AbstractAcceptanceTest {

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private MockMvc mockMvc;

    @Test
    void basicFlowStartingFromScratch() throws Exception {
        createDeveloper("dev-0");
        createDeveloper("dev-1");
        createDeveloper("dev-2");
        createDeveloper("dev-3");

        List<Long> developersToIncludeInCombinations = getDeveloperIdsFor(List.of(
                "dev-0",
                "dev-1",
                "dev-2"
        ));

        createStream("stream-a");
        createStream("stream-b");

        List<Long> streamsToIncludeInCombinations = getStreamIdsFor(List.of(
                "stream-a",
                "stream-b"
        ));

        final List<ScoredCombination> scoredCombinations = calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations);

        assertThat(scoredCombinations).isNotEmpty();

        final ScoredCombination bestCombination = scoredCombinations.getFirst();

        assertThat(bestCombination.combination().stream()
                .flatMap(ps -> ps.developers().stream())
                .map(DeveloperInfo::displayName)
                .toList())
                .containsExactlyInAnyOrder("dev-0", "dev-1", "dev-2");

        assertThat(bestCombination.combination().stream()
                .map(PairStream::stream)
                .map(StreamInfo::displayName)
                .toList())
                .containsExactlyInAnyOrder("stream-a", "stream-b");

        final List<CombinationEvent> startingEvents = findAllCombinationEvents();

        assertThat(startingEvents).isEmpty();

        saveCombinationEventFor(LocalDate.of(2024, 4, 27), bestCombination.combination());

        final List<CombinationEvent> allCombinationEvents = findAllCombinationEvents();

        assertThat(allCombinationEvents).hasSize(1);

        final CombinationEvent savedEvent = allCombinationEvents.getFirst();

        assertThat(savedEvent.date()).isEqualTo(LocalDate.of(2024, 4, 27));
        assertThat(savedEvent.combination()).hasSize(2);
        assertThat(savedEvent.combination()).isEqualTo(bestCombination.combination());
    }

    @Test
    void exampleOfSecondDay() throws Exception {
        basicFlowStartingFromScratch();

        final CombinationEvent yesterdayEvent = findAllCombinationEvents().getFirst();

        List<Long> developersToIncludeInCombinations = getDeveloperIdsFor(List.of(
                "dev-0",
                "dev-1",
                "dev-2"
        ));

        List<Long> streamsToIncludeInCombinations = getStreamIdsFor(List.of(
                "stream-a",
                "stream-b"
        ));

        final List<ScoredCombination> scoredCombinations = calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations);

        assertThat(scoredCombinations).isNotEmpty();

        final List<PairStream> todayCombination = scoredCombinations.getFirst().combination();

        final List<PairStream> yesterdayCombination = yesterdayEvent.combination();

        assertThat(todayCombination)
                .isNotEqualTo(yesterdayCombination);


        saveCombinationEventFor(LocalDate.of(2024, 4, 28), todayCombination);

        final List<CombinationEvent> allCombinationEvents = findAllCombinationEvents();

        assertThat(allCombinationEvents).hasSize(2);

        final CombinationEvent savedEvent = allCombinationEvents.getFirst();

        assertThat(savedEvent.date()).isEqualTo(LocalDate.of(2024, 4, 28));
        assertThat(savedEvent.combination()).hasSize(2);

        deleteCombinationEvent(savedEvent.id());

        final List<CombinationEvent> allCombinationEventsAfterDelete = findAllCombinationEvents();

        assertThat(allCombinationEventsAfterDelete).hasSize(1);
        assertThat(allCombinationEventsAfterDelete).containsExactly(yesterdayEvent);
    }

    private void createDeveloper(String developerName) throws Exception {
        mockMvc.perform(post("/api/v1/developers")
                .contentType(MediaType.APPLICATION_JSON)
                .content("""
                        {
                          "name": "%s"
                        }""".formatted(developerName))

        ).andExpect(status().isCreated());
    }

    private List<Long> getDeveloperIdsFor(List<String> developerNames) throws Exception {
        final MvcResult result = mockMvc.perform(get("/api/v1/developers"))
                .andExpect(status().isOk())
                .andReturn();
        final String responseBody = result.getResponse()
                .getContentAsString();

        final List<Developer> developers = objectMapper.readValue(responseBody, new TypeReference<>() {
        });

        return developers.stream()
                .filter(developer -> developerNames.contains(developer.name()))
                .map(Developer::id)
                .toList();
    }

    private void createStream(String streamName) throws Exception {
        mockMvc.perform(post("/api/v1/streams")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "name": "%s"
                                }""".formatted(streamName))
                )
                .andExpect(status().isCreated());
    }

    private List<Long> getStreamIdsFor(List<String> streamNames) throws Exception {
        final String responseBody = mockMvc.perform(get("/api/v1/streams"))
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        final List<Stream> streams = objectMapper.readValue(responseBody, new TypeReference<>() {
        });

        return streams.stream()
                .filter(stream -> streamNames.contains(stream.name()))
                .map(Stream::id)
                .toList();
    }

    private List<ScoredCombination> calculateCombinations(List<Long> developerIds, List<Long> streamIds) throws Exception {
        final CalculateInputDto input = new CalculateInputDto(developerIds, streamIds);

        final String responseBody = mockMvc.perform(post("/api/v1/combinations/calculate")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(input))
                )
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        return objectMapper.readValue(responseBody, new TypeReference<>() {
        });
    }

    private void saveCombinationEventFor(LocalDate date, List<PairStream> combination) throws Exception {
        List<SaveCombinationEventDto.PairStreamByIds> combinationByIds = combination.stream()
                .map(ps -> {
                    List<Long> developerIds = ps.developers().stream()
                            .map(DeveloperInfo::id)
                            .toList();
                    return new SaveCombinationEventDto.PairStreamByIds(developerIds, ps.stream().id());
                })
                .toList();

        final SaveCombinationEventDto dto = new SaveCombinationEventDto(date, combinationByIds);


        mockMvc.perform(post("/api/v1/combinations/event")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(dto))
                )
                .andExpect(status().isCreated());
    }

    private List<CombinationEvent> findAllCombinationEvents() throws Exception {
        final String responseBody = mockMvc.perform(get("/api/v1/combinations/event"))
                .andExpect(status().isOk())
                .andReturn()
                .getResponse().getContentAsString();
        return objectMapper.readValue(responseBody, new TypeReference<>() {
        });
    }

    private void deleteCombinationEvent(long id) throws Exception {
        mockMvc.perform(delete("/api/v1/combinations/event/{id}", id))
                .andExpect(status().isNoContent());
    }
}

File: backend/src/test/kotlin/dev/coldhands/pair/stairs/backend/AbstractAcceptanceTest.kt
package dev.coldhands.pair.stairs.backend

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import dev.coldhands.pair.stairs.backend.domain.*
import dev.coldhands.pair.stairs.backend.domain.developer.Developer
import dev.coldhands.pair.stairs.backend.domain.developer.DeveloperStats
import dev.coldhands.pair.stairs.backend.domain.stream.StreamStats
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CalculateInputDto
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.SaveCombinationEventDto
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.collections.shouldNotBeEmpty
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldStartWith
import jakarta.transaction.Transactional
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.MvcResult
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
abstract class AbstractAcceptanceTest(
    private val objectMapper: ObjectMapper,
    private val mockMvc: MockMvc,
    private val dbUrl: String
) {

    protected abstract fun getExpectedDbUrlPrefix(): String

    @Test
    fun checkDbUrl() {
        dbUrl.shouldStartWith(getExpectedDbUrlPrefix())
    }

    @Test
    fun basicFlowStartingFromScratch() {
        val dev0Id = createDeveloper("dev-0")
        createDeveloper("dev-1")
        createDeveloper("dev-2")
        createDeveloper("dev-3")

        val developersToIncludeInCombinations = getDeveloperIdsFor(
            listOf("dev-0", "dev-1", "dev-2")
        )

        val streamAId = createStream("stream-a")
        createStream("stream-b")

        val streamsToIncludeInCombinations = getStreamIdsFor(
            listOf("stream-a", "stream-b")
        )

        val scoredCombinations =
            calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations)

        scoredCombinations.shouldNotBeEmpty()

        val bestCombination = scoredCombinations.first()

        bestCombination.combination()
            .flatMap { it.developers() }
            .map(DeveloperInfo::displayName)
            .shouldContainExactlyInAnyOrder("dev-0", "dev-1", "dev-2")

        bestCombination.combination()
            .map { it.stream().displayName() }
            .shouldContainExactlyInAnyOrder("stream-a", "stream-b")

        val startingEvents = findAllCombinationEvents()
        startingEvents.shouldBeEmpty()

        val today = LocalDate.of(2024, 4, 27)
        saveCombinationEventFor(today, bestCombination.combination())

        val allCombinationEvents = findAllCombinationEvents()
        allCombinationEvents shouldHaveSize 1

        val savedEvent = allCombinationEvents.first()
        savedEvent.date() shouldBe today
        savedEvent.combination() shouldHaveSize 2
        savedEvent.combination() shouldBe bestCombination.combination()

        val developerStats = getDeveloperStatsBetween(dev0Id, today, today)
        developerStats.developerStats shouldHaveSize 4
        developerStats.streamStats shouldHaveSize 2

        val streamStats = getStreamStatsBetween(streamAId, today, today)
        streamStats.developerStats shouldHaveSize 4
    }

    @Test
    fun exampleOfSecondDay() {
        basicFlowStartingFromScratch()

        val yesterdayEvent = findAllCombinationEvents().first()

        val developersToIncludeInCombinations = getDeveloperIdsFor(
            listOf("dev-0", "dev-1", "dev-2")
        )
        val streamsToIncludeInCombinations = getStreamIdsFor(
            listOf("stream-a", "stream-b")
        )

        val scoredCombinations =
            calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations)
        scoredCombinations.shouldNotBeEmpty()

        val todayCombination = scoredCombinations.first().combination()
        val yesterdayCombination = yesterdayEvent.combination()

        todayCombination shouldNotBe yesterdayCombination

        val tomorrow = LocalDate.of(2024, 4, 28)
        saveCombinationEventFor(tomorrow, todayCombination)

        val allCombinationEvents = findAllCombinationEvents()
        allCombinationEvents shouldHaveSize 2

        val savedEvent = allCombinationEvents.first()
        savedEvent.date() shouldBe tomorrow
        savedEvent.combination() shouldHaveSize 2

        deleteCombinationEvent(savedEvent.id())

        val allCombinationEventsAfterDelete = findAllCombinationEvents()
        allCombinationEventsAfterDelete shouldHaveSize 1
        allCombinationEventsAfterDelete shouldBe listOf(yesterdayEvent)
    }

    private fun createDeveloper(developerName: String): DeveloperId {
        val result: MvcResult = mockMvc.perform(
            post("/api/v1/developers")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                      "name": "$developerName"
                    }
                    """.trimIndent()
                )
        ).andExpect(status().isCreated)
            .andReturn()

        val responseBody = result.response.contentAsString
        val developer: Developer = objectMapper.readValue(responseBody)
        return developer.id
    }

    private fun getDeveloperIdsFor(developerNames: List<String>): List<DeveloperId> {
        val result = mockMvc.perform(get("/api/v1/developers"))
            .andExpect(status().isOk)
            .andReturn()

        val responseBody = result.response.contentAsString
        val developers: List<Developer> = objectMapper.readValue(responseBody)

        return developers
            .filter { it.name in developerNames }
            .map(Developer::id)
    }

    private fun createStream(streamName: String): Long {
        val result = mockMvc.perform(
            post("/api/v1/streams")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                      "name": "$streamName"
                    }
                    """.trimIndent()
                )
        )
            .andExpect(status().isCreated)
            .andReturn()

        val responseBody = result.response.contentAsString
        val stream: Stream = objectMapper.readValue(responseBody)
        return stream.id()
    }

    private fun getStreamIdsFor(streamNames: List<String>): List<Long> {
        val responseBody = mockMvc.perform(get("/api/v1/streams"))
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        val streams: List<Stream> = objectMapper.readValue(responseBody)

        return streams
            .filter { it.name() in streamNames }
            .map(Stream::id)
    }

    private fun calculateCombinations(developerIds: List<DeveloperId>, streamIds: List<Long>): List<ScoredCombination> {
        val input = CalculateInputDto(developerIds, streamIds)

        val responseBody = mockMvc.perform(
            post("/api/v1/combinations/calculate")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(input))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        return objectMapper.readValue(responseBody)
    }

    private fun saveCombinationEventFor(date: LocalDate, combination: List<PairStream>) {
        val combinationByIds: List<SaveCombinationEventDto.PairStreamByIds> =
            combination.map { ps ->
                val developerIds = ps.developers().map(DeveloperInfo::id)
                // todo just use DeveloperId directly once DeveloperInfo is kotlin-ed
                SaveCombinationEventDto.PairStreamByIds(developerIds.map { DeveloperId(it) }, ps.stream().id())
            }

        val dto = SaveCombinationEventDto(date, combinationByIds)

        mockMvc.perform(
            post("/api/v1/combinations/event")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(dto))
        )
            .andExpect(status().isCreated)
    }

    private fun findAllCombinationEvents(): List<CombinationEvent> {
        val responseBody = mockMvc.perform(get("/api/v1/combinations/event"))
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        return objectMapper.readValue(responseBody)
    }

    private fun deleteCombinationEvent(id: Long) {
        mockMvc.perform(delete("/api/v1/combinations/event/{id}", id))
            .andExpect(status().isNoContent)
    }

    private fun getDeveloperStatsBetween(
        developerId: DeveloperId,
        startDate: LocalDate,
        endDate: LocalDate
    ): DeveloperStats {
        val responseBody = mockMvc.perform(
            get("/api/v1/developers/{id}/stats", developerId.value)
                .queryParam("startDate", startDate.format(DateTimeFormatter.ISO_DATE))
                .queryParam("endDate", endDate.format(DateTimeFormatter.ISO_DATE))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        return objectMapper.readValue(responseBody)
    }

    private fun getStreamStatsBetween(
        streamId: Long,
        startDate: LocalDate,
        endDate: LocalDate
    ): StreamStats {
        val responseBody = mockMvc.perform(
            get("/api/v1/streams/{id}/stats", streamId)
                .queryParam("startDate", startDate.format(DateTimeFormatter.ISO_DATE))
                .queryParam("endDate", endDate.format(DateTimeFormatter.ISO_DATE))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        return objectMapper.readValue(responseBody)
    }
}
